src/Str.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace OL\String;
+
+use OL\String\Contracts\Str as StrContract;
+
+/**
+ * Class Str.
+ *
+ * Basic manipulation of strings.
+ */
+class Str implements StrContract
+{
+    /**
+     * @var string Raw string content
+     */
+    protected $payload;
+
+    /**
+     * Str constructor.
+     *
+     * The constructor will enforce string type, since it can also receive
+     * another Str object (self) as it can be casted to string.
+     *
+     * @param string|StrContract $payload The actual string
+     */
+    public function __construct(string $payload)
+    {
+        // casts the payload as string and them assign on class scope
+        $this->payload = (string) $payload;
+    }
+
+    /**
+     * Str objects should should be able to be cast as string.
+     *
+     * @return string The actual string
+     */
+    public function __toString()
+    {
+        return $this->contents();
+    }
+
+    /**
+     * Public method to retrieve the raw string contents of an object.
+     *
+     * @return string
+     */
+    public function contents()
+    {
+        return $this->payload;
+    }
+
+    /**
+     * Replace one or more entries from the given object with a given value
+     *
+     * @param array|string $search What to search for
+     * @param string $replace Replace with?
+     *
+     * @return StrContract Object instance with the replaced text.
+     */
+    public function replace($search, string $replace)
```

Comment:
I think, at least for the "primitive components" (string, number and maybe array), we should use immutable objects. Immutability achieves a defensive way of programming by preventing objects to change their state, which is a preference for some developers. Also, it's the same implementation of the native PHP functions, making the transition from the existing code to PHP-OL more "natural".

I also think this should be discussed in project-level.